File: swift-bridge/Sources/ModelIOBridge/PhysicallyPlausibleLight.swift
import Foundation
import ModelIO

private func mdl_physically_plausible_light_info(_ light: MDLPhysicallyPlausibleLight) -> [String: Any] {
    [
        "light_type": light.lightType.rawValue,
        "color_space": light.colorSpace,
        "color": mdl_color_components(light.color) as Any,
        "lumens": light.lumens,
        "inner_cone_angle": light.innerConeAngle,
        "outer_cone_angle": light.outerConeAngle,
        "attenuation_start_distance": light.attenuationStartDistance,
        "attenuation_end_distance": light.attenuationEndDistance,
    ]
}

private func mdl_area_light_info(_ light: MDLAreaLight) -> [String: Any] {
    var info = mdl_physically_plausible_light_info(light)
    info["area_radius"] = light.areaRadius
    info["super_elliptic_power"] = [light.superEllipticPower.x, light.superEllipticPower.y]
    info["aspect"] = light.aspect
    return info
}

private func mdl_photometric_light_info(_ light: MDLPhotometricLight) -> [String: Any] {
    var info = mdl_physically_plausible_light_info(light)
    info["spherical_harmonics_level"] = light.sphericalHarmonicsLevel
    info["spherical_harmonics_coefficients_length"] = light.sphericalHarmonicsCoefficients?.count ?? 0
    info["has_light_cube_map"] = light.lightCubeMap != nil
    return info
}

@_cdecl("mdl_physically_plausible_light_new")
public func mdl_physically_plausible_light_new(
    _ outLight: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let outLight else {
            throw ModelIOBridgeError.invalidArgument("missing output light pointer")
        }
        outLight.pointee = mdl_retain(MDLPhysicallyPlausibleLight())
    }
}

@_cdecl("mdl_physically_plausible_light_info_json")
public func mdl_physically_plausible_light_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return nil }
    return mdl_string(mdl_json_string(from: mdl_physically_plausible_light_info(light)) ?? "{}")
}

@_cdecl("mdl_physically_plausible_light_set_color_temperature")
public func mdl_physically_plausible_light_set_color_temperature(_ handle: UnsafeMutableRawPointer?, _ temperature: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.setColorByTemperature(temperature)
}

@_cdecl("mdl_physically_plausible_light_set_color")
public func mdl_physically_plausible_light_set_color(
    _ handle: UnsafeMutableRawPointer?,
    _ red: Float,
    _ green: Float,
    _ blue: Float,
    _ alpha: Float
) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.color = mdl_color(red, green, blue, alpha)
}

@_cdecl("mdl_physically_plausible_light_set_lumens")
public func mdl_physically_plausible_light_set_lumens(_ handle: UnsafeMutableRawPointer?, _ lumens: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.lumens = lumens
}

@_cdecl("mdl_physically_plausible_light_set_inner_cone_angle")
public func mdl_physically_plausible_light_set_inner_cone_angle(_ handle: UnsafeMutableRawPointer?, _ angle: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.innerConeAngle = angle
}

@_cdecl("mdl_physically_plausible_light_set_outer_cone_angle")
public func mdl_physically_plausible_light_set_outer_cone_angle(_ handle: UnsafeMutableRawPointer?, _ angle: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.outerConeAngle = angle
}

@_cdecl("mdl_physically_plausible_light_set_attenuation_start_distance")
public func mdl_physically_plausible_light_set_attenuation_start_distance(_ handle: UnsafeMutableRawPointer?, _ distance: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.attenuationStartDistance = distance
}

@_cdecl("mdl_physically_plausible_light_set_attenuation_end_distance")
public func mdl_physically_plausible_light_set_attenuation_end_distance(_ handle: UnsafeMutableRawPointer?, _ distance: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLPhysicallyPlausibleLight else { return }
    light.attenuationEndDistance = distance
}

@_cdecl("mdl_area_light_new")
public func mdl_area_light_new(
    _ outLight: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let outLight else {
            throw ModelIOBridgeError.invalidArgument("missing output area light pointer")
        }
        outLight.pointee = mdl_retain(MDLAreaLight())
    }
}

@_cdecl("mdl_area_light_info_json")
public func mdl_area_light_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let light = mdl_borrow_object(handle) as? MDLAreaLight else { return nil }
    return mdl_string(mdl_json_string(from: mdl_area_light_info(light)) ?? "{}")
}

@_cdecl("mdl_area_light_set_area_radius")
public func mdl_area_light_set_area_radius(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLAreaLight else { return }
    light.areaRadius = value
}

@_cdecl("mdl_area_light_set_super_elliptic_power")
public func mdl_area_light_set_super_elliptic_power(_ handle: UnsafeMutableRawPointer?, _ x: Float, _ y: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLAreaLight else { return }
    light.superEllipticPower = SIMD2<Float>(x, y)
}

@_cdecl("mdl_area_light_set_aspect")
public func mdl_area_light_set_aspect(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let light = mdl_borrow_object(handle) as? MDLAreaLight else { return }
    light.aspect = value
}

@_cdecl("mdl_photometric_light_new")
public func mdl_photometric_light_new(
    _ outLight: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let outLight else {
            throw ModelIOBridgeError.invalidArgument("missing output photometric light pointer")
        }
        outLight.pointee = mdl_retain(MDLPhotometricLight())
    }
}

@_cdecl("mdl_photometric_light_new_with_ies_profile")
public func mdl_photometric_light_new_with_ies_profile(
    _ path: UnsafePointer<CChar>?,
    _ outLight: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let path, let outLight else {
            throw ModelIOBridgeError.invalidArgument("missing IES path or output photometric light pointer")
        }
        let url = URL(fileURLWithPath: String(cString: path))
        guard let light = MDLPhotometricLight(iesProfile: url) else {
            throw ModelIOBridgeError.nullResult("MDLPhotometricLight IES initializer returned nil")
        }
        outLight.pointee = mdl_retain(light)
    }
}

@_cdecl("mdl_photometric_light_info_json")
public func mdl_photometric_light_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let light = mdl_borrow_object(handle) as? MDLPhotometricLight else { return nil }
    return mdl_string(mdl_json_string(from: mdl_photometric_light_info(light)) ?? "{}")
}

@_cdecl("mdl_photometric_light_generate_spherical_harmonics_from_light")
public func mdl_photometric_light_generate_spherical_harmonics_from_light(_ handle: UnsafeMutableRawPointer?, _ level: UInt64) {
    guard let light = mdl_borrow_object(handle) as? MDLPhotometricLight else { return }
    light.generateSphericalHarmonics(fromLight: Int(level))
}

@_cdecl("mdl_photometric_light_generate_cubemap_from_light")
public func mdl_photometric_light_generate_cubemap_from_light(_ handle: UnsafeMutableRawPointer?, _ textureSize: UInt64) {
    guard let light = mdl_borrow_object(handle) as? MDLPhotometricLight else { return }
    light.generateCubemap(fromLight: Int(textureSize))
}

@_cdecl("mdl_photometric_light_generate_texture")
public func mdl_photometric_light_generate_texture(_ handle: UnsafeMutableRawPointer?, _ textureSize: UInt64) -> UnsafeMutableRawPointer? {
    guard let light = mdl_borrow_object(handle) as? MDLPhotometricLight else { return nil }
    return mdl_retain(light.generateTexture(Int(textureSize)))
}

@_cdecl("mdl_photometric_light_light_cube_map")
public func mdl_photometric_light_light_cube_map(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let light = mdl_borrow_object(handle) as? MDLPhotometricLight,
          let texture = light.lightCubeMap
    else {
        return nil
    }
    return mdl_retain(texture)
}

File: src/physically_plausible_light.rs
use std::path::Path;
use std::ptr;

use crate::error::Result;
use crate::ffi;
use crate::handle::ObjectHandle;
use crate::light::Light;
use crate::object::Object;
use crate::texture::Texture;
use crate::types::{AreaLightInfo, PhotometricLightInfo, PhysicallyPlausibleLightInfo};
use crate::util::{parse_json, path_to_c_string, required_handle};

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O physically plausible light counterpart.
pub struct PhysicallyPlausibleLight {
    handle: ObjectHandle,
}

impl PhysicallyPlausibleLight {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O physically plausible light counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O physically plausible light counterpart.
    pub fn new() -> Result<Self> {
        let mut out_light = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        let status =
            // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
            unsafe { ffi::mdl_physically_plausible_light_new(&mut out_light, &mut out_error) };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_light,
            "MDLPhysicallyPlausibleLight",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn info(&self) -> Result<PhysicallyPlausibleLightInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_physically_plausible_light_info_json(self.handle.as_ptr()) },
            "MDLPhysicallyPlausibleLight",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_color_temperature(&self, temperature: f32) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_physically_plausible_light_set_color_temperature(
                self.handle.as_ptr(),
                temperature,
            );
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_color(&self, color: [f32; 4]) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_physically_plausible_light_set_color(
                self.handle.as_ptr(),
                color[0],
                color[1],
                color[2],
                color[3],
            );
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_lumens(&self, lumens: f32) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_physically_plausible_light_set_lumens(self.handle.as_ptr(), lumens) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_inner_cone_angle(&self, angle: f32) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_physically_plausible_light_set_inner_cone_angle(self.handle.as_ptr(), angle);
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_outer_cone_angle(&self, angle: f32) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_physically_plausible_light_set_outer_cone_angle(self.handle.as_ptr(), angle);
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_attenuation_start_distance(&self, distance: f32) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_physically_plausible_light_set_attenuation_start_distance(
                self.handle.as_ptr(),
                distance,
            );
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn set_attenuation_end_distance(&self, distance: f32) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_physically_plausible_light_set_attenuation_end_distance(
                self.handle.as_ptr(),
                distance,
            );
        };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn as_light(&self) -> Light {
        Light::from_handle(self.handle.clone())
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O physically plausible light counterpart.
    pub fn as_object(&self) -> Object {
        Object::from_handle(self.handle.clone())
    }
}

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O area light counterpart.
pub struct AreaLight {
    handle: ObjectHandle,
}

impl AreaLight {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O area light counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O area light counterpart.
    pub fn new() -> Result<Self> {
        let mut out_light = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
        let status = unsafe { ffi::mdl_area_light_new(&mut out_light, &mut out_error) };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_light,
            "MDLAreaLight",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn info(&self) -> Result<AreaLightInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_area_light_info_json(self.handle.as_ptr()) },
            "MDLAreaLight",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn set_area_radius(&self, value: f32) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_area_light_set_area_radius(self.handle.as_ptr(), value) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn set_super_elliptic_power(&self, value: [f32; 2]) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_area_light_set_super_elliptic_power(self.handle.as_ptr(), value[0], value[1]);
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn set_aspect(&self, value: f32) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_area_light_set_aspect(self.handle.as_ptr(), value) };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn as_physically_plausible_light(&self) -> PhysicallyPlausibleLight {
        PhysicallyPlausibleLight::from_handle(self.handle.clone())
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn as_light(&self) -> Light {
        Light::from_handle(self.handle.clone())
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O area light counterpart.
    pub fn as_object(&self) -> Object {
        Object::from_handle(self.handle.clone())
    }
}

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O photometric light counterpart.
pub struct PhotometricLight {
    handle: ObjectHandle,
}

impl PhotometricLight {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O photometric light counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O photometric light counterpart.
    pub fn new() -> Result<Self> {
        let mut out_light = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
        let status = unsafe { ffi::mdl_photometric_light_new(&mut out_light, &mut out_error) };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_light,
            "MDLPhotometricLight",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn from_ies_profile(path: impl AsRef<Path>) -> Result<Self> {
        let path = path_to_c_string(path.as_ref())?;
        let mut out_light = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: The unsafe operation is valid in this context.
        let status = unsafe {
            ffi::mdl_photometric_light_new_with_ies_profile(
                path.as_ptr(),
                &mut out_light,
                &mut out_error,
            )
        };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_light,
            "MDLPhotometricLight",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn info(&self) -> Result<PhotometricLightInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_photometric_light_info_json(self.handle.as_ptr()) },
            "MDLPhotometricLight",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn generate_spherical_harmonics_from_light(&self, level: usize) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_photometric_light_generate_spherical_harmonics_from_light(
                self.handle.as_ptr(),
                level as u64,
            );
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn generate_cubemap_from_light(&self, texture_size: usize) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_photometric_light_generate_cubemap_from_light(
                self.handle.as_ptr(),
                texture_size as u64,
            );
        };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn generate_texture(&self, texture_size: usize) -> Option<Texture> {
        // SAFETY: The unsafe operation is valid in this context.
        let ptr = unsafe {
            ffi::mdl_photometric_light_generate_texture(self.handle.as_ptr(), texture_size as u64)
        };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Texture::from_handle)
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn light_cube_map(&self) -> Option<Texture> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_photometric_light_light_cube_map(self.handle.as_ptr()) };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Texture::from_handle)
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn as_physically_plausible_light(&self) -> PhysicallyPlausibleLight {
        PhysicallyPlausibleLight::from_handle(self.handle.clone())
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn as_light(&self) -> Light {
        Light::from_handle(self.handle.clone())
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O photometric light counterpart.
    pub fn as_object(&self) -> Object {
        Object::from_handle(self.handle.clone())
    }
}
